src/object/cli_obj.c

```diff
@@ -983,6 +985,113 @@ struct shard_update_args {
 	daos_sg_list_t		*sgls;
 };
 
+/* prepare the bulk handle(s) for obj request */
+static int
+obj_bulk_prep(daos_sg_list_t *sgls, unsigned int nr, bool bulk_bind,
+	      crt_bulk_perm_t bulk_perm, tse_task_t *task,
+	      struct obj_auxi_args *obj_auxi)
+{
+	crt_bulk_t	*bulks;
+	int		 i = 0;
+	int		 rc = 0;
+
+	D_ASSERTF(nr >= 1, "invalid nr %d.\n", nr);
+	D_ALLOC_ARRAY(bulks, nr);
```

Comment:
now the obj_auxi is actually in the task tse priviate space's dtp_buf

Reply:
"if we do server io dispatch, does this save us anything?"
I think it does not save bulk creation for server-io dispatch, but avoid create duplicated bulk handle for client io dispatch.

Reply:
TBH, we will probably never use client io dispatch, especially after we discussed distribute transaction over beijing meeting. So I wonder whether we should complicate the layer here. since with this patch, it will call shard layer with both sgl and bulk?

Reply:
client IO dispatch is still kept (I don't remember we ever decided to remove it?), and this patch can avoid recreate bulk handle for IO retry case even for srv dispatch. 
Actually for IO retry, it possibly cause some problems if destroy+recreate bulk handle as original code in timeout+retry (possibly cause server-side bulk error if old destroyed bulk handle used by server in race condition).

Reply:
ah, can we fix it without changing bulk creation?

Reply:
I thought client io dispatch will only be kept for testing purpose, if I understand correctly.

Reply:
"can we fix it without changing bulk creation?"
I think it also can be refined in shard layer that don't destroy bulk handle when the shard task will be retried and reuse it when retry. Just that moving it in obj layer seems clearer to me.
This patch is not emergent, I'm fine to delay it later together with client IO stack refining for EC.